### DISCUSSION TITLE
flag language and syntax features into stable or unstable versions & language edition field in v.mod

### DISCUSSION THREAD

**Discussion by gcxfd** [vlang/v#14630] [Ideas]
I really appreciate this view of the V language

https://github.com/vlang/v/blob/master/CONTRIBUTING.md

> I tried to make the code of the compiler and vlib as simple and readable as possible. One of V's goals is to be open to developers with different levels of experience in compiler development. Compilers don't need to be black boxes full of magic that only few people understand.

Programming languages are supposed to be able to iterate quickly, with small steps of trial and error, and everyone can participate in the development of new features of the programming language.

However, new syntax and language features, which can have a variety of problems, are difficult to use with confidence in major projects without being marked.

I think v-language should flag language and syntax features into stable and unstable versions, as rust does.

https://doc.rust-lang.org/beta/unstable-book/language-features/ffi-const.html
<img width="1275" alt="image" src="https://user-images.githubusercontent.com/69608605/154882449-e215b2ad-c5a0-4edd-8a8f-19604c3549ae.png">

People who write toys can try the latest language features. And people who intend to use the v-language for major projects online will avoid using features that are not yet stable.

If you want to use unstable features, you need to show the declaration in the entry file.

It is possible that some language features may affect existing code, so I think it is also important to support flag language versions like rust, so as not to split the vlang ecosystem

[https://doc.rust-lang.org/cargo/reference/manifest.html#the-edition-field](https://doc.rust-lang.org/cargo/reference/manifest.html#the-edition-field)

[https://doc.rust-lang.org/edition-guide/editions/index.html](https://doc.rust-lang.org/edition-guide/editions/index.html)

### [The  `edition`  field](https://doc.rust-lang.org/cargo/reference/manifest.html#the-edition-field)

> The  `edition`  key is an optional key that affects which  [Rust Edition](https://doc.rust-lang.org/edition-guide/index.html)  your package is compiled with. Setting the  `edition`  key in  `[package]`  will affect all targets/crates in the package, including test suites, benchmarks, binaries, examples, etc.
>
> [package]
> edition = '2021'` 
>
> Most manifests have the  `edition`  field filled in automatically by  [`cargo new`](https://doc.rust-lang.org/cargo/commands/cargo-new.html)  with the latest stable edition. By default  `cargo new`  creates a manifest with the 2021 edition currently.
>
> If the  `edition`  field is not present in  `Cargo.toml`, then the 2015 edition is assumed for backwards compatibility. Note that all manifests created with  [`cargo new`](https://doc.rust-lang.org/cargo/commands/cargo-new.html)  will not use this historical fallback because they will have  `edition`  explicitly specified to a newer value.

###  [What are Editions?](https://doc.rust-lang.org/edition-guide/editions/index.html#what-are-editions)

The release of Rust 1.0 established ["stability without stagnation"](https://blog.rust-lang.org/2014/10/30/Stability.html) as a core Rust deliverable. Ever since the 1.0 release, the rule for Rust has been that once a feature has been released on stable, we are committed to supporting that feature for all future releases.

There are times, however, when it is useful to be able to make small changes to the language that are not backwards compatible. The most obvious example is introducing a new keyword, which would invalidate variables with the same name. For example, the first version of Rust did not have the async and await keywords. Suddenly changing those words to keywords in a later version would've broken code like let async = 1;.

Editions are the mechanism we use to solve this problem. When we want to release a feature that would otherwise be backwards incompatible, we do so as part of a new Rust edition. Editions are opt-in, and so existing crates do not see these changes until they explicitly migrate over to the new edition. This means that even the latest version of Rust will still not treat async as a keyword, unless edition 2018 or later is chosen. This choice is made per crate [as part of its Cargo.toml](https://doc.rust-lang.org/cargo/reference/manifest.html#the-edition-field). New crates created by cargo new are always configured to use the latest stable edition.

[Editions do not split the ecosystem](https://doc.rust-lang.org/edition-guide/editions/index.html#editions-do-not-split-the-ecosystem) The most important rule for editions is that crates in one edition can interoperate seamlessly with crates compiled in other editions. This ensures that the decision to migrate to a newer edition is a "private one" that the crate can make without affecting others.

The requirement for crate interoperability implies some limits on the kinds of changes that we can make in an edition. In general, changes that occur in an edition tend to be "skin deep". All Rust code, regardless of edition, is ultimately compiled to the same internal representation within the compiler.

[Edition migration is easy and largely automated](https://doc.rust-lang.org/edition-guide/editions/index.html#edition-migration-is-easy-and-largely-automated) Our goal is to make it easy for crates to upgrade to a new edition. When we release a new edition, we also provide [tooling to automate the migration](https://doc.rust-lang.org/cargo/commands/cargo-fix.html). It makes minor changes to your code necessary to make it compatible with the new edition. For example, when migrating to Rust 2018, it changes anything named async to use the equivalent [raw identifier syntax](https://doc.rust-lang.org/rust-by-example/compatibility/raw_identifiers.html)[](https://doc.rust-lang.org/edition-guide/introduction.html): r#async.

The automated migrations are not necessarily perfect: there might be some corner cases where manual changes are still required. The tooling tries hard to avoid changes to semantics that could affect the correctness or performance of the code.

In addition to tooling, we also maintain this Edition Migration Guide that covers the changes that are part of an edition. This guide describes each change and gives pointers to where you can learn more about it. It also covers any corner cases or details you should be aware of. This guide serves both as an overview of the edition and as a quick troubleshooting reference if you encounter problems with the automated tooling.

**Comment by JalonSolov** [comment]
That's all great, and will likely be something that should be done when V reaches v1.0.  In the meantime, V is still _ALPHA_ software, at v0.2.
